File: csharp/flower-field/FlowerField.cs
public static class FlowerField
{
    public static string[] Annotate(string[] input)
    {
        var flowers = FindFlowers(input);
        return input
        .Select((line, i) => new string(line.Select((ch, j) => ch == '*' ? '*' : CountNeighboringFlowers(i, j, flowers)).ToArray()))
        .ToArray();
    }

    private static List<(int row, int col)> FindFlowers(string[] input) =>
    input.SelectMany((line, row) => line.Select((ch, col) => (ch, row, col)).Where(x => x.ch == '*').Select(x => (x.row, x.col))).ToList();

    private static char CountNeighboringFlowers(int i, int j, List<(int i, int j)> flowers)
    {
        var neighbors = new List<(int i, int j)>
        {   // 8 neighboring positions, the missing field is the position itself
            (i - 1, j - 1), (i - 1, j), (i - 1, j + 1),
            (i    , j - 1),             (i    , j + 1),
            (i + 1, j - 1), (i + 1, j), (i + 1, j + 1)
        };
        var count = neighbors.Count(n => flowers.Contains(n));
        var count_char = (char)('0' + count);
        return count == 0 ? ' ' : count_char;
    }
}

File: fsharp/flower-field/FlowerField.fs
module FlowerField

let annotate (input : string list) = 
    let flowers = 
        [ for r, line in List.indexed input do
          for c, ch in Seq.indexed line do
          if ch = '*' then yield (r, c) ] |> Set.ofList
    let neighbors (r, c) =
        [ for dr in -1 .. 1 do
          for dc in -1 .. 1 do
          if not (dr = 0 && dc = 0) then
              yield (r + dr, c + dc) ]
    input |> List.mapi (fun r line ->
        line |> Seq.mapi (fun c ch ->
            if ch = '*' then '*'
            else
                let count =
                    neighbors (r, c)
                    |> List.filter (fun pos -> flowers |> Set.contains pos)
                    |> List.length
                if count = 0 then ' ' else char (count + int '0')
        ) |> Seq.toArray |> System.String
    )